providers/implementations/encode_decode/ml_kem_codecs.c

```diff
@@ -0,0 +1,559 @@
+                      const char *propq)
+{
+    OSSL_LIB_CTX *libctx = PROV_LIBCTX_OF(provctx);
+    const ML_KEM_VINFO *v;
+    const ML_KEM_CINFO *c;
+    ML_KEM_PKCS8_PREF *vp8_alloc = NULL, *vp8_slot;
+    const ML_KEM_PKCS8_INFO *vp8;
+    ML_KEM_KEY *key = NULL, *ret = NULL;
+    PKCS8_PRIV_KEY_INFO *p8inf = NULL;
+    const uint8_t *buf, *pos;
+    const X509_ALGOR *alg = NULL;
+    const char *formats;
+    int len, ptype;
+    uint32_t magic;
+    uint16_t seed_magic;
+
+    /* Which ML-KEM variant? */
+    if ((v = ossl_ml_kem_get_vinfo(evp_type)) == NULL
+        || (c = ml_kem_get_cinfo(evp_type)) == NULL)
+        return 0;
+
+    /* Extract the key OID and any parameters. */
+    if ((p8inf = d2i_PKCS8_PRIV_KEY_INFO(NULL, &prvenc, prvlen)) == NULL)
+        return 0;
+    /* Shortest prefix is 4 bytes: seq tag/len  + octet string tag/len */
+    if (!PKCS8_pkey_get0(NULL, &buf, &len, &alg, p8inf))
+        goto end;
+    /* Bail out early if this is some other key type. */
+    if (OBJ_obj2nid(alg->algorithm) != evp_type)
+        goto end;
+
+    /* Get the list of enabled decoders. Their order is not important here. */
+    formats = ossl_prov_ctx_get_param(
+        provctx, OSSL_PKEY_PARAM_ML_KEM_INPUT_FORMATS, NULL);
+    vp8_slot = vp8_alloc = vp8_order(v->algorithm_name, c->pkcs8_info,
+                                     "input", formats);
+    if (vp8_alloc == NULL)
+        goto end;
+
+    /* Parameters must be absent. */
+    X509_ALGOR_get0(NULL, &ptype, NULL, alg);
+    if (ptype != V_ASN1_UNDEF) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_UNEXPECTED_KEY_PARAMETERS,
+                       "unexpected parameters with a PKCS#8 %s private key",
+                       v->algorithm_name);
+        goto end;
+    }
+    if ((ossl_ssize_t)len < (ossl_ssize_t)sizeof(magic))
+        goto end;
+
+    /* Find the matching p8 info slot, this has the expected length. */
+    pos = OPENSSL_load_u32_be(&magic, buf);
+    for (vp8_slot = vp8_alloc; vp8_slot->vp8_entry != NULL; ++vp8_slot) {
+        if (magic == vp8_slot->vp8_entry->p8_magic
+            && len == (ossl_ssize_t)vp8_slot->vp8_entry->p8_bytes)
+            break;
+    }
+    if ((vp8 = vp8_slot->vp8_entry) == NULL)
+        goto end;
+
+    if (vp8->seed_offset > 0) {
+        /* Check |seed| tag/len, if not subsumed by |magic|. */
+        if (pos == buf + vp8->seed_offset - 2) {
+            pos = OPENSSL_load_u16_be(&seed_magic, pos);
+            if (seed_magic != vp8->seed_magic)
+                goto end;
+        } else if (pos != buf + vp8->seed_offset) {
+            goto end;
+        }
+        pos += ML_KEM_SEED_BYTES;
+    }
+    if (vp8->priv_offset > 0) {
+        /* Check |priv| tag/len */
+        if (pos == buf + vp8->priv_offset - 4) {
+            pos = OPENSSL_load_u32_be(&magic, pos);
+            if (magic != vp8->priv_magic)
+                goto end;
+        } else if (pos != buf + vp8->priv_offset) {
+            goto end;
```

Comment:
put a comment here of which format type this applies to..

Reply:
This applies to any format that has a private key component. What do you want the comment to say?